### DISCUSSION TITLE
Bug with validating array of ints in query params

### DISCUSSION THREAD

**Discussion by EvgeniyMoiseevUnifonic** [fastapi/fastapi#10457] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import datetime
from typing import Union, Self, Annotated

from fastapi import FastAPI, Query, Depends
from pydantic import BaseModel, model_validator, field_validator, Field

app = FastAPI()


class GetUserLogsRequestInScheme(BaseModel):
    start_date: datetime.date
    end_date: datetime.date
    agent_ids: list[int] | None = Field(
        Query([]),  # hack to make it work properly with query params (https://github.com/tiangolo/fastapi/issues/4445#issuecomment-1117632409)
        max_length=100
    )

    @model_validator(mode="after")
    def validate_dates(self) -> Self:
        if self.start_date > self.end_date:
            raise ValueError("start_date can't be greater then end_date")
        return self

    @field_validator("agent_ids", mode="before")
    @classmethod
    def validate_agent_ids(cls, v: Union[list[int], str, int]) -> list[int]:
        # this method is never called
        if isinstance(v, str):
            v = [int(str_) for str_ in v.split(",")]
        return v


@app.get("/logs")
async def get_logs(query_params: Annotated[GetUserLogsRequestInScheme, Depends()]) -> None:
    return
```

### Description

I'm trying to achieve quite a simple thing - validate an array of the integers in a query string, but it looks like `@filed_validator` with a `mode="before"` argument doesn't work as expected.

Steps to reproduce:
Open the browser and call http://localhost:8000/logs?start_date=2023-11-01&end_date=2023-11-02&agent_ids=1,2

Expected behavior - handler returns `null`

Actual behavior - handler returns validation error
```{
    "detail": [
        {
            "type": "int_parsing",
            "loc": [
                "query",
                "agent_ids",
                0
            ],
            "msg": "Input should be a valid integer, unable to parse string as an integer",
            "input": "1,2",
            "url": "https://errors.pydantic.dev/2.4/v/int_parsing"
        }
    ]
}```

### Operating System

macOS

### Operating System Details

_No response_

### FastAPI Version

0.103.2

### Pydantic Version

2.4.2

### Python Version

3.11.5

### Additional Context

_No response_

**Comment by ebottos94** [comment]
Simply this is not the correct way in a url to use list query parameters. 

Use this instead :

http://localhost:8000/logs?start_date=2023-11-01&end_date=2023-11-02&agent_ids=1&agent_ids=2

**Reply by hasansezertasan** [reply]
I think the url was generated by Swagger UI. I setup this one and couldn't see a "+" button for adding items, just "text" field...

**Reply by EvgeniyMoiseevUnifonic** [reply]
1. I'm in circumstances where I can't change the frontend code.
2. I would like to see some standard that specifies it is incorrect because I don't know any.
3. The bug is still there - `field_validator` with mode `before` is not working when pydantic model is applied to query params

**Reply by ebottos94** [reply]
@hasansezertasan I think that there is a bug in Swagger UI for list, and so in docs is not testable. Anyway the url that @EvgeniyMoiseevUnifonic posted is not correct for a list, agent_ids=1,2 in query parameters is interpreted as a string "1,2" for this is raised a validation error.

**Reply by EvgeniyMoiseevUnifonic** [reply]
@ebottos94 I still don't understand why it is not correct for a list

**Reply by hasansezertasan** [reply]
Is this fixed? It might be because of the Swagger UI which is based on JSON Schema definition.

**Reply by EvgeniyMoiseevUnifonic** [reply]
@hasansezertasan, my issue is not about swagger (while it exists), my issue is about the standard validation mechanism that's not working when applied to query params

**Reply by EvgeniyMoiseevUnifonic** [reply]
@tiangolo can I convert this to a bug?

**Reply by croumegous** [reply]
I have the exact same issue with pydantic v0.109.2 and pydantic 2.6.0
```python
class User(BaseModel):
    fields: List[str] = Field([])
    
    @field_validator("fields", mode="before")
    @classmethod
    def validate_fields(cls, value: str) -> List[str]:
        if not value:            
            return []
        return value.split(",")
```

```python
User(fields="name,age").model_dump()
# {'fields': ['name', 'age']}
```
Pydantic returns the expected results.

But with fastapi custom query_params it fails:

```python
app = FastAPI()

# Define a simple endpoint
@app.get("/")
def read_root(args: Annotated[User, Depends()]) -> User:
    return args
```

```python
# http://127.0.0.1:8000/?fields=test,name
{
    "fields": []
}
```
I know that with fastapi we should call the endpoint like that `http://127.0.0.1:8000/?fields=test&fields=name` but I can't change the frontend calls this is why I try to use this custom "before" validator in pydantic to fix the input format. But fastapi use the default value

**Reply by croumegous** [reply]
BTW if I use :
```python
class User(BaseModel):
    fields: str = Field("")
    
    @field_validator("fields")
    @classmethod
    def validate_fields(cls, value: str) -> List[str]:
        if not value:            
            return value
        return value.split(",")
```
It works well

**Comment by estebanx64** [comment, marked as answer]
Hi there.

As many others have mentioned before, you are handling a list of query params in a wrong way. The documentation is clear here how to handle them, then it's not a bug.
https://fastapi.tiangolo.com/tutorial/query-params-str-validations/#query-parameter-list-multiple-values

and just following your needs, here's a few ways to make it happen.

FastAPI as a web framework is just processing your url as a string and pydantic with the type annotations is waiting for a List of int, it is not going to work, so the workaround would be tell pydantic is going to get a str like this

```python

import datetime
from typing import Annotated
from fastapi import FastAPI, Depends
from pydantic import BaseModel, field_validator, Field

app = FastAPI()

class GetUserLogsRequestInScheme(BaseModel):
    start_date: datetime.date = Field(..., example="2023-01-01")
    end_date: datetime.date = Field(..., example="2023-01-02")
    agent_ids: str = Field(default="", example="1,2,3")

    @field_validator('agent_ids')
    def split_str_to_int_list(cls, value: str) -> list[int]:
        if isinstance(value, str):
            return [int(x) for x in value.split(',')]
        return value


@app.get("/logs")
async def get_logs(query_params: Annotated[GetUserLogsRequestInScheme, Depends()]) -> dict:
    # Use query_params as needed here
    return {"message": "Success"}

```

another solution would be parsing your query params with just simple python functions and not using pydantic like this.
```python
from typing import List, Optional
from fastapi import FastAPI, Query, Depends
from pydantic import BaseModel, validator
import datetime

app = FastAPI()

def parse_agent_ids(agent_ids: Optional[str] = Query(None)):
    # This function will parse the agent_ids from a comma-separated string to a list of integers.
    return [int(agent_id) for agent_id in agent_ids.split(",")] if agent_ids else []

class GetUserLogsRequestInScheme(BaseModel):
    start_date: datetime.date
    end_date: datetime.date

    @validator('end_date')
    def check_dates(cls, value, values, **kwargs):
        if 'start_date' in values and value < values['start_date']:
            raise ValueError("end_date must be after start_date")
        return value

@app.get("/logs")
async def get_logs(start_date: datetime.date, end_date: datetime.date, agent_ids: List[int] = Depends(parse_agent_ids)):
    # Now agent_ids should correctly contain the list of integers, even if passed as a comma-separated string.
    # Do something with the validated data here.
    return {"message": "Success", "agent_ids": agent_ids}

```

**Reply by CompadreP** [reply]
@estebanx64 of course there can be 1001 workarounds for this problem, but I'm pretty sure we can name it a bug because it's not working as expected. Why should we use hacks/workarounds if the proper way to handle enormous inputs is using `@field_validator` with `mode=before` and it's just not working as expected in this case?

**Reply by CompadreP** [reply]
@estebanx64 please write an answer to my comment or I'll have to open the same discussion again
